README.md
# eReuse.org Workbench Desktop
## What does this program do?
eReuse.org Workbench Desktop is a desktop cross-platform application which intends to extract information about the hardware of computer devices and deliver a snapshot.
## How it works?
![Workbench Desktop](./diagram/Workbench_Desktop.png)
## Dependencies
In order to execute this program properly you have to install the SmartmonTools version 7 or superior program and the dmidecode package (only in Linux).
### Linux
It is available in most Linux systems. Depending on your distribution you have to introduce the next command in a terminal:
- Debian/Ubuntu/Linux Mint and Debian based distributions: **$ sudo apt-get install smartmontools dmidecode**
- Arch Linux and Arch based distributions: **$ sudo pacman -S smartmontools dmidecode**
- Fedora/Centos/OpenSuse and Red Hat derivates: **$ sudo yum install smartmontools dmidecode**
- Solus: **$ sudo eopkg it smartmontools dmidecode**
### Windows
The program is already included in the installation process, you only have to execute the installer and follow the instructions.
### MacOS
If you do not have installed brew in MacOS, open a terminal and type in the next command: **$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"** \
Then type the following in your terminal: **$ brew install smartmontools**
## Building
In order to build the different releases for the objective operating system, the usage of the PyInstaller library is mandatory.
### Linux
- Execute: **pyinstaller --windowed -F --icon="path to .ico" -n "workbench" workbench.py**
### Windows
- Execute: **pyinstaller --windowed -F --uac-admin --icon="path to .ico" -n "workbench" workbench.py**
- There is a script writen that 
### MacOS
- Execute: **pyinstaller --windowed -F --icon="path to .icns" -n "workbench" workbench.py**
## Installation
### Linux
- Extract the workbench.tar.gz file: **$ tar -xvf workbench.tar.gz**
### Windows
- Execute the installer and follow the steps
- After that the program should be placed in Program Files (x86)\\eReuse.org\\Workbench
### MacOS
- Extract the workbench.tar.gz file: **$ tar -xvf workbench.tar.gz**
## Usage
### Linux
- Execute the program with admin privileges: **$ sudo ./workbench**
### Windows
- Execute the program placed in **Program Files (x86)\\eReuse.org\\Workbench**
### MacOS
- Execute the program with admin privileges: **$ sudo ./workbench**


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## master
  [0.0.0]

## testing
  [1.0.0b4]

## [1.0.0b4]
- [remove] Clean up of all non-essential files
- [addend] Add example folder
- [addend] Add GPLV3 License
- [changes] Formal changes to set up for 1.0.0b4


workbench_desktop/api.py
import json

import requests
import semver


def send_snapshot(domain, snapshot):
    URL_upload = "https://" + domain + "/lives/"
    post_headers = {'Content-type': 'application/json'}
    snapshot_json = json.dumps(snapshot, indent=2)
    try:
        response = requests.post(URL_upload, data=snapshot_json, headers=post_headers)
        if response.status_code == 201:
            return 0
        else:
            return -1
    except:
        return -1


def get_license(domain, version, language):
    URL_license = "http://" + domain + "/licences/"
    try:
        response = requests.get(URL_license)
        if response.status_code == 200:
            licenses = json.loads(response.json())
            for license in licenses:
                if semver.compare(license["WorkbenchDesktopVersion"], version) == 0 and license["Language"] == language:
                    last_license = license
            return last_license
        else:
            return "Can't retrieve valid license"
    except:
        return "Can't communicate with the license server"


workbench_desktop/linux.py
import json
import platform
import subprocess
import time
import uuid

from dmiparser import DmiParser


class Linux:

    def __init__(self):
        self.snapshot = dict()
        self.desktop = None

    def do_snapshot(self, version):
        self.snapshot['uuid'] = str(uuid.uuid4())
        self.snapshot['type'] = 'Live'
        self.snapshot['software'] = 'WorkbenchDesktop'
        self.snapshot['version'] = version
        # self.snapshot['os'] = platform.platform(terse=True).split('-')[0] + " " + \
        #                      platform.platform(terse=True).split('-')[1]

        device_dict = self.device()

        components = list()
        # self.motherboard(components)
        # self.processor(components)
        # self.memory(components)
        self.disk(components)
        # self.graphics(components)
        self.network_interface(components)
        # self.power_supply(components)

        self.snapshot['device'] = device_dict
        self.snapshot['components'] = components

        self.snapshot['licence_version'] = '0.0.0'
        return self.snapshot

    # obtain System related information
    def device(self):
        device_info = dict()

        cmd = ["dmidecode -t 1"]
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        output, errors = proc.communicate()
        proc.wait()

        if proc.returncode >= 0:
            output_json = json.loads(str(DmiParser(output.decode('utf8'))))
            try:
                device_info['serialNumber'] = output_json[0]['props']['Serial Number']['values'][0]
            except:
                device_info['serialNumber'] = None
            try:
                device_info['sku'] = output_json[0]['props']['SKU Number']['values'][0]
            except:
                device_info['sku'] = None
            try:
                device_info['manufacturer'] = output_json[0]['props']['Manufacturer']['values'][0]
            except:
                device_info['manufacturer'] = None
            try:
                device_info['model'] = output_json[0]['props']['Version']['values'][0]
            except:
                device_info['model'] = output_json[0]['props']['Product Name']['values'][0]

        else:
            device_info['serialNumber'] = -1

        cmd = ["dmidecode -t 3"]
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        output, errors = proc.communicate()
        proc.wait()

        if proc.returncode >= 0:
            output_json = json.loads(str(DmiParser(output.decode('utf8'))))
            try:
                device_info['chassis'] = output_json[0]['props']['Type']['values'][0]
            except:
                device_info['chassis'] = None
            if device_info['chassis'].find('book') != -1:
                self.desktop = False
                device_info['type'] = 'Laptop'
                device_info['chassis'] = 'Netbook'
            else:
                self.desktop = True
                device_info['type'] = 'Desktop'
                device_info['chassis'] = 'Tower'

        return device_info

    # obtain Motherboard related information
    @staticmethod
    def motherboard(components):
        motherboard = dict()

        cmd = ["dmidecode -t 2"]
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        output, errors = proc.communicate()
        proc.wait()

        if proc.returncode >= 0:
            output_json = json.loads(str(DmiParser(output.decode('utf8'))))

            motherboard['type'] = 'Motherboard'
            motherboard['manufacturer'] = output_json[0]['props']['Manufacturer']['values'][0]
            motherboard['model'] = output_json[0]['props']['Product Name']['values'][0]
            motherboard['serialNumber'] = output_json[0]['props']['Serial Number']['values'][0]
        else:
            motherboard['type'] = 'Motherboard'
            motherboard["serialNumber"] = -1

        components.append(motherboard)

    # obtain CPU related information
    @staticmethod
    def processor(components):
        cpu = dict()

        cmd = ["dmidecode -t 4"]
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        output, errors = proc.communicate()
        proc.wait()

        if proc.returncode >= 0:
            output_json = json.loads(str(DmiParser(output.decode('utf8'))))

            cpu['type'] = 'Processor'
            cpu['manufacturer'] = output_json[0]['props']['Manufacturer']['values'][0]
            cpu['model'] = output_json[0]['props']['Version']['values'][0]
            cpu['address'] = 64
            cpu['speed'] = float(output_json[0]['props']['Current Speed']['values'][0].split(' ')[0]) / 1000
            cpu['cores'] = int(output_json[0]['props']['Core Count']['values'][0])
            cpu['threads'] = int(output_json[0]['props']['Thread Count']['values'][0])
            cpu['serialNumber'] = None
        else:
            cpu['type'] = 'Processor'
            cpu['serialNumber'] = -1

        components.append(cpu)

    # obtain RAM information
    @staticmethod
    def memory(components):
        cmd = ["dmidecode -t 17"]
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        output, errors = proc.communicate()
        proc.wait()
        output_json = json.loads(str(DmiParser(output.decode('utf8'))))

        if proc.returncode >= 0:
            for dimm in output_json:
                if dimm['props']['Size']['values'][0] != 'No Module Installed':
                    ram = {'type': 'RamModule',
                           'speed': int(dimm['props']['Speed']['values'][0].split(' ')[0]),
                           'interface': dimm['props']['Type']['values'][0],
                           'format': dimm['props']['Form Factor']['values'][0],
                           'serialNumber': dimm['props']['Serial Number']['values'][0],
                           'manufacturer': dimm['props']['Manufacturer']['values'][0]}

                    if int(dimm['props']['Size']['values'][0].split(' ')[0]) > 128:
                        ram['size'] = int(dimm['props']['Size']['values'][0].split(' ')[0])
                    else:
                        ram['size'] = int(dimm['props']['Size']['values'][0].split(' ')[0]) * 1024
                    components.append(ram)
        else:
            ram = {'type': 'RamModule', 'serialNumber': -1}
            components.append(ram)

    # obtain Disks information
    @staticmethod
    def disk(components):
        cmd = ["lsblk -J"]
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        output, errors = proc.communicate()
        proc.wait()
        output_json = json.loads(str(output.decode('utf8')))

        if proc.returncode >= 0:
            for disk in output_json['blockdevices']:
                if disk['name'].find('loop') == -1:
                    cmd = ["smartctl --all -j /dev/" + disk['name']]
                    start_time_smart = time.time()
                    proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
                    output, errors = proc.communicate()
                    proc.wait()

                    if proc.returncode >= 0:
                        smart_json = json.loads(str(output.decode('utf8')))
                        elapsed_time_smart = time.time() - start_time_smart

                        disk_json = {}

                        if smart_json.get("smart_family"):
                            if smart_json["model_family"].find("SSD") == -1:
                                disk_json["type"] = "HardDrive"
                            else:
                                disk_json["type"] = "SolidStateDrive"
                        else:
                            if smart_json["model_name"].find("SSD") == -1:
                                disk_json["type"] = "HardDrive"
                            else:
                                disk_json["type"] = "SolidStateDrive"

                            disk_json['model'] = smart_json['model_name']
                            disk_json['serialNumber'] = smart_json['serial_number']
                            disk_json['size'] = smart_json['user_capacity']['bytes'] / 1024 / 1024

                            actions = []
                            try:
                                smart = {'type': 'TestDataStorage', 'powerCycleCount': smart_json['power_cycle_count'],
                                         'length': 'Short',
                                         'status': 'Completed without error', 'elapsed': elapsed_time_smart}
                            except:
                                smart = {'type': 'TestDataStorage', 'powerCycleCount': 0,
                                         'length': 'Short', 'status': "Can't evaluate SMART",
                                         'elapsed': elapsed_time_smart}
                            try:
                                smart['lifetime'] = smart_json['ata_smart_self_test_log']['standard']['table'][0][
                                    "lifetime_hours"]
                            except:
                                smart['lifetime'] = 0
                            actions.append(smart)

                            benchmark = {}
                            start_time_benchmark_hdd = time.time()
                            cmd = ["dd if=/dev/" + disk['name'] + " of=/dev/null bs=1M count=256 oflag=dsync"]
                            proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
                            output, errors = proc.communicate()
                            proc.wait()
                            output = output.decode('utf8')
                            if proc.returncode >= 0:
                                value = float(output.split()[-2].replace(',', '.'))
                                if value < 20.0:
                                    value = value * 1000.0
                                benchmark['readSpeed'] = value
                            else:
                                benchmark['readSpeed'] = None

                            cmd = ["dd if=/dev/" + disk['name'] + " of=/dev/" + disk[
                                'name'] + "bs=1M count=256 oflag=dsync"]
                            proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
                            output, errors = proc.communicate()
                            proc.wait()
                            output = output.decode('utf8')
                            if proc.returncode >= 0:
                                value = float(output.split()[-2].replace(',', '.'))
                                if value < 20.0:
                                    value = value * 1000.0
                                benchmark['writeSpeed'] = value
                            else:
                                benchmark['writeSpeed'] = None

                            elapsed_time_benchmark_hdd = time.time() - start_time_benchmark_hdd
                            benchmark['elapsed'] = elapsed_time_benchmark_hdd
                            benchmark['type'] = 'BenchmarkDataStorage'
                            actions.append(benchmark)
                            disk_json['actions'] = actions

                            components.append(disk_json)
                    else:
                        disk_json = {'type': 'DataStorage', 'serialNumber': -1}
                        components.append(disk_json)

        else:
            disk_json = {'type': 'DataStorage', 'serialNumber': -1}
            components.append(disk_json)

    # obtain GPU information
    @staticmethod
    def graphics(components):
        gpu = {}

        cmd = ["lspci | grep VGA | cut -d "":"" -f3"]
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        output, errors = proc.communicate()
        proc.wait()

        if proc.returncode >= 0:
            gpu['type'] = 'GraphicCard'
            if output.decode('utf8').strip().find('NVIDIA') != -1:
                gpu['manufacturer'] = 'NVIDIA Corporation'
            elif output.decode('utf8').strip().find('AMD') != -1:
                gpu['manufacturer'] = 'AMD'
            else:
                gpu['manufacturer'] = 'Intel Corp.'
            gpu['model'] = output.decode('utf8').strip()
            gpu['memory'] = None
            gpu['serialNumber'] = None
        else:
            gpu['type'] = 'GraphicCard'
            gpu['serialNumber'] = -1

        components.append(gpu)

    # obtain Network Information
    @staticmethod
    def network_interface(components):
        cmd = ["ip -j a"]
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        output, errors = proc.communicate()
        proc.wait()

        if proc.returncode >= 0:
            output_json = json.loads(str(output.decode('utf8')))
            for nic in output_json:
                if nic['ifname'].find('vir') == -1 and nic['ifname'].find('docker') == -1 and nic['ifname'].find(
                        'lo') == -1:
                    interface = {'type': 'NetworkAdapter', 'serialNumber': nic['address'], 'wireless': False}

                    components.append(interface)
        else:
            interface = {'type': 'NetworkAdapter', 'serialNumber': -1}
            components.append(interface)

    # obtain Power supply information (only laptop)
    def power_supply(self, components):
        if not self.desktop:
            cmd = ["dmidecode -t 22"]
            proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
            output, errors = proc.communicate()
            proc.wait()

            if proc.returncode >= 0:
                output_json = json.loads(str(DmiParser(output.decode('utf8'))))
                for battery in output_json:
                    battery_json = {'type': 'Battery', 'manufacturer': battery['props']['Manufacturer']['values'][0],
                                    'size': int(battery['props']['Design Capacity']['values'][0].split(' ')[0]),
                                    'model': None, 'serialNumber': None}
                    components.append(battery_json)
            else:
                battery_json = {'type': 'Battery', 'serialNumber': -1}
                components.append(battery_json)


workbench_desktop/windows.py
import json
import os
import platform
import subprocess
import time
import uuid

import utils

os.environ["COMSPEC"] = 'powershell'


class Windows:

    def __init__(self):
        self.snapshot = dict()
        self.desktop = True

    def do_snapshot(self, version):
        self.snapshot['uuid'] = str(uuid.uuid4())
        self.snapshot['type'] = 'Live'
        self.snapshot['software'] = 'WorkbenchDesktop'
        self.snapshot['version'] = version
        self.snapshot['os'] = platform.platform(terse=True).split('-')[0] + " " + \
                              platform.platform(terse=True).split('-')[1]

        device_dict = self.device()

        components = list()

        self.motherboard(components)
        self.processor(components)
        self.memory(components)
        self.disk(components)
        self.graphics(components)
        self.network_interface(components)
        self.power_supply(components)

        self.snapshot['device'] = device_dict
        self.snapshot['components'] = components

        return self.snapshot

    # obtain System related information
    def device(self):
        device_info = dict()

        cmd = ['Get-CimInstance -ClassName Win32_systemenclosure | select SerialNumber | ConvertTo-JSON']
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                                stdin=subprocess.PIPE)
        output, errors = proc.communicate()
        proc.wait()
        if proc.returncode >= 0:
            output_json = json.loads(output)
            device_info['serialNumber'] = output_json['SerialNumber']
        else:
            device_info['serialNumber'] = -1

        cmd = [
            'Get-CimInstance -ClassName Win32_ComputerSystem | select Manufacturer, SystemFamily, SystemSKUNumber | '
            'ConvertTo-JSON']
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                                stdin=subprocess.PIPE)
        output, errors = proc.communicate()
        proc.wait()

        if proc.returncode >= 0:
            output_json = json.loads(output)
            device_info['manufacturer'] = output_json['Manufacturer']
            device_info['sku'] = output_json['SystemSKUNumber']
            device_info['model'] = output_json['SystemFamily']
        else:
            device_info['serialNumber'] = -1

        cmd = ['Get-CimInstance -ClassName Win32_PortableBattery | select Availability | ConvertTo-JSON']
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                                stdin=subprocess.PIPE)
        output, errors = proc.communicate()
        proc.wait()

        if proc.returncode >= 0:
            if len(output.decode('utf8')) == 0:
                self.desktop = True
                device_info['type'] = 'Desktop'
                device_info['chassis'] = 'Tower'
            else:
                self.desktop = False
                device_info['type'] = 'Laptop'
                device_info['chassis'] = 'Netbook'

        return device_info

    # obtain Motherboard related information
    @staticmethod
    def motherboard(components):
        motherboard = dict()

        cmd = [
            'Get-CimInstance -ClassName Win32_BaseBoard | select Manufacturer, Product, SerialNumber | ConvertTo-JSON']
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                                stdin=subprocess.PIPE)
        output, errors = proc.communicate()
        proc.wait()

        if proc.returncode >= 0:
            output_json = json.loads(output)

            motherboard['type'] = 'Motherboard'
            motherboard['manufacturer'] = output_json['Manufacturer']
            motherboard['model'] = output_json['Product']
            motherboard['serialNumber'] = output_json['SerialNumber']
        else:
            motherboard['type'] = 'Motherboard'
            motherboard['serialNumber'] = -1

        components.append(motherboard)

    # obtain CPU related information
    @staticmethod
    def processor(components):
        cpu = dict()

        cmd = [
            'Get-CimInstance -ClassName Win32_Processor | select Name, MaxClockSpeed, AddressWidth, NumberOfCores, '
            'NumberOfLogicalProcessors | ConvertTo-JSON']
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                                stdin=subprocess.PIPE)
        output, errors = proc.communicate()
        proc.wait()

        if proc.returncode >= 0:
            output_json = json.loads(output)

            cpu['type'] = 'Processor'

            if output_json['Name'].find('Intel') != -1:
                cpu['manufacturer'] = 'Intel Corp.'
            elif output_json['Name'].find('AMD') != -1:
                cpu['manufacturer'] = 'AMD'
            else:
                cpu['manufacturer'] = 'other'

            cpu['model'] = output_json['Name'].strip()
            cpu['speed'] = output_json['MaxClockSpeed'] / 1000
            cpu['address'] = output_json['AddressWidth']
            cpu['cores'] = output_json['NumberOfCores']
            cpu['threads'] = output_json['NumberOfLogicalProcessors']
            cpu['serialNumber'] = None
        else:
            cpu['type'] = 'Processor'
            cpu['serialNumber'] = -1

        components.append(cpu)

    # obtain RAM information
    def memory(self, components):
        cmd = [
            'Get-CimInstance -ClassName Win32_PhysicalMemory | select Manufacturer, Capacity, ConfiguredClockSpeed, '
            'PartNumber | ConvertTo-JSON']
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                                stdin=subprocess.PIPE)
        output, errors = proc.communicate()
        proc.wait()

        if proc.returncode >= 0:
            output_json = json.loads(output)
            if type(output_json) == list:
                for dimm in output_json:
                    ram_json = {'type': 'RamModule'}
                    if self.desktop:
                        ram_json['format'] = 'DIMM'
                    else:
                        ram_json['format'] = 'SODIMM'
                    ram_json['size'] = dimm['Capacity'] / 1024 / 1024
                    ram_json['manufacturer'] = dimm['Manufacturer']
                    ram_json['speed'] = dimm['ConfiguredClockSpeed']
                    ram_json['interface'] = None
                    ram_json['serialNumber'] = None
                    components.append(ram_json)
            else:
                ram_json = {'type': 'RamModule'}
                if self.desktop:
                    ram_json['format'] = 'DIMM'
                else:
                    ram_json['format'] = 'SODIMM'
                ram_json['size'] = output_json['Capacity'] / 1024 / 1024
                ram_json['manufacturer'] = output_json['Manufacturer']
                ram_json['speed'] = output_json['ConfiguredClockSpeed']
                ram_json['interface'] = None
                ram_json['serialNumber'] = None
                components.append(ram_json)

        else:
            ram_json = {'type': 'RamModule', 'serialNumber': -1}
            components.append(ram_json)

    # obtain Disks information
    @staticmethod
    def disk(components):
        cmd = ['smartctl -j --scan']
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                                stdin=subprocess.PIPE)
        output, errors = proc.communicate()
        proc.wait()

        if proc.returncode >= 0:
            output_json = json.loads(output)
            count_disk = 0
            for disk in output_json['devices']:
                if disk['type'] == 'ata' and disk['name'].find('/dev/sd') != -1:
                    cmd = [
                        'Get-Disk -Number ' + str(count_disk) + ' | select Model, Size, SerialNumber | ConvertTo-JSON']
                    proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                                            stdin=subprocess.PIPE)
                    output, errors = proc.communicate()
                    proc.wait()
                    disk_info_json = json.loads(output)

                    disk_json = {'model': disk_info_json['Model'],
                                 'serialNumber': disk_info_json['SerialNumber'].strip(), 'interface': 'ATA'}
                    size_gb = int(disk_info_json['Size'])
                    size_gb = int(size_gb / 1000 / 1000)
                    if size_gb != 0:
                        disk_json['size'] = size_gb

                    cmd = ['smartctl -j --all ' + disk['name']]
                    start_time_smart = time.time()
                    proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                                            stdin=subprocess.PIPE)
                    output, errors = proc.communicate()
                    proc.wait()
                    smart_json = json.loads(output)
                    actions = []
                    if proc.returncode == 0:
                        elapsed_time_smart = time.time() - start_time_smart
                        if smart_json.get("smart_family"):
                            if smart_json["model_family"].find("SSD") == -1:
                                disk_json["type"] = "HardDrive"
                            else:
                                disk_json["type"] = "SolidStateDrive"
                        else:
                            if smart_json["model_name"].find("SSD") == -1:
                                disk_json["type"] = "HardDrive"
                            else:
                                disk_json["type"] = "SolidStateDrive"
                        try:
                            smart = {'type': 'TestDataStorage', 'powerCycleCount': smart_json['power_cycle_count'],
                                     'length': 'Short',
                                     'status': 'Completed without error', 'elapsed': elapsed_time_smart}
                        except:
                            smart = {'type': 'TestDataStorage', 'powerCycleCount': 0,
                                     'length': 'Short', 'status': "Can't evaluate SMART", 'elapsed': elapsed_time_smart}
                        try:
                            smart['lifetime'] = smart_json['ata_smart_self_test_log']['standard']['table'][0][
                                "lifetime_hours"]
                        except:
                            smart['lifetime'] = 0
                        actions.append(smart)

                        benchmark = {}
                        start_time_benchmark_hdd = time.time()
                        results = utils.disk_test()
                        elapsed_time_benchmark_hdd = time.time() - start_time_benchmark_hdd
                        benchmark['type'] = 'BenchmarkDataStorage'
                        benchmark['writeSpeed'] = (1048576 / float(results['write'])) / 1000000
                        benchmark['readSpeed'] = (1048576 / float(results['read'])) / 10000000
                        benchmark['elapsed'] = elapsed_time_benchmark_hdd
                        actions.append(benchmark)
                        disk_json['actions'] = actions
                        components.append(disk_json)
                        count_disk = count_disk + 1
                        os.remove("testfile")
                    else:
                        smart = {'type': 'TestDataStorage',
                                 'length': 'Short', 'status': 'Completed with error'}
                        actions.append(smart)
                        benchmark = {}
                        start_time_benchmark_hdd = time.time()
                        results = utils.disk_test()
                        elapsed_time_benchmark_hdd = time.time() - start_time_benchmark_hdd
                        benchmark['type'] = 'BenchmarkDataStorage'
                        benchmark['writeSpeed'] = (1048576 / float(results['write'])) / 1000000
                        benchmark['readSpeed'] = (1048576 / float(results['read'])) / 10000000
                        benchmark['elapsed'] = elapsed_time_benchmark_hdd
                        actions.append(benchmark)
                        disk_json['actions'] = actions
                        components.append(disk_json)
                        count_disk = count_disk + 1
                        os.remove("testfile")
        else:
            disk_json = {'type': 'DataStorage', 'serialNumber': -1}
            components.append(disk_json)

    # obtain GPU information
    @staticmethod
    def graphics(components):
        gpu = {}
        cmd = [
            'Get-CimInstance -ClassName Win32_VideoController | select Name, AdapterCompatibility, AdapterRAM | '
            'ConvertTo-JSON']
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                                stdin=subprocess.PIPE)
        output, errors = proc.communicate()
        proc.wait()

        if proc.returncode >= 0:
            output_json = json.loads(output)

            gpu['type'] = 'GraphicCard'
            gpu['manufacturer'] = output_json['AdapterCompatibility']
            gpu['model'] = output_json['Name']
            gpu['memory'] = int(output_json['AdapterRAM'] / 1024 / 1024) * 2
            gpu['serialNumber'] = None
        else:
            gpu['type'] = 'GraphicCard'
            gpu['serialNumber'] = -1
        components.append(gpu)

    # obtain Network Information
    @staticmethod
    def network_interface(components):
        cmd = [
            'Get-CimInstance -ClassName Win32_NetworkAdapter | select Name, Speed, MACAddress, Caption, Manufacturer, '
            'ProductName | ConvertTo-JSON']
        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                                stdin=subprocess.PIPE)
        output, errors = proc.communicate()
        proc.wait()

        if proc.returncode >= 0:
            output_json = json.loads(output)
            for NIC in output_json:
                if NIC['Speed'] is not None and NIC['Caption'].find('Bluetooth') == -1 and NIC['Name'].find(
                        'Virtual') == -1 and NIC['Name'].find('TAP') == -1 and NIC['ProductName'].find('Tunnel') == -1:
                    interface_json = {'type': 'NetworkAdapter', 'manufacturer': NIC['Manufacturer'],
                                      'model': NIC['ProductName'], 'serialNumber': NIC['MACAddress'], 'wireless': False}
                    if NIC['Speed'] / 1000000 > 10000:
                        interface_json['speed'] = int(NIC['Speed'] / 10000000000000000)
                    else:
                        interface_json['speed'] = int(NIC['Speed'] / 1000000)
                    components.append(interface_json)
        else:
            interface_json = {'type': 'NetworkAdapter', 'serialNumber': -1}
            components.append(interface_json)

    # obtain Power supply information (only laptop)
    def power_supply(self, components):
        if not self.desktop:
            cmd = [
                'Get-CimInstance -ClassName Win32_PortableBattery | select Manufacturer, Name, DesignCapacity '
                '| ConvertTo-JSON']
            proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                                    stdin=subprocess.PIPE)
            output, errors = proc.communicate()
            proc.wait()
            if proc.returncode >= 0:
                output_json = json.loads(output)
                if isinstance(output_json, dict):
                    battery_json = {'type': 'Battery', 'manufacturer': output_json['Manufacturer'],
                                    'model': output_json['Name'], 'technology': 'LiIon', 'wireless': False,
                                    'size': output_json['DesignCapacity']}
                    components.append(battery_json)
                else:
                    for battery in output_json:
                        battery_json = {'type': 'Battery', 'manufacturer': battery['Manufacturer'],
                                        'model': battery['Name'], 'size': battery['DesignCapacity'],
                                        'serialNumber': None}
                        components.append(battery_json)
            else:
                battery_json = {'type': 'Battery', 'serialNumber': -1}
                components.append(battery_json)
